File: backend/app/Models/Report.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Report extends Model
{
    protected $fillable = [
        'user_id',
        'user_name',
        'incident_type',
        'description',
        'latitude',
        'longitude',
        'user_agent',
        'verification_count',
        'status',
        'report_time',
    ];

    /**
     * Get the votes for the report.
     */
    public function votes(): HasMany
    {
        return $this->hasMany(ReportVote::class);
    }

    /**
     * Get the user who created the report.
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Check if a user has voted on this report.
     */
    public function hasUserVoted(?int $userId, ?string $deviceId = null): bool
    {
        if ($userId) {
            return $this->votes()->where('user_id', $userId)->exists();
        }

        if ($deviceId) {
            return $this->votes()->where('device_id', $deviceId)->exists();
        }

        return false;
    }

    /**
     * Get user's vote type on this report.
     */
    public function getUserVoteType(?int $userId, ?string $deviceId = null): ?string
    {
        $vote = null;

        if ($userId) {
            $vote = $this->votes()->where('user_id', $userId)->first();
        } elseif ($deviceId) {
            $vote = $this->votes()->where('device_id', $deviceId)->first();
        }

        return $vote?->vote_type;
    }
}

File: mobile-app/app/src/main/java/com/example/fr/data/models/Report.kt
package com.example.fr.data.models

import kotlinx.serialization.Serializable

@Serializable
data class Report(
    val id: Int,
    val user_id: Int? = null,
    val user_name: String,
    val incident_type: String,
    val description: String,
    val latitude: Double,
    val longitude: Double,
    val user_agent: String? = null,
    val verification_count: Int = 0,
    val status: String? = null,
    val report_time: String? = null,
    val created_at: String? = null,
    val updated_at: String? = null,
    // Vote status from backend
    val user_vote: String? = null,
    val has_voted: Boolean = false
)

@Serializable
data class ReportRequest(
    val user_name: String,
    val incident_type: String,
    val description: String,
    val latitude: Double,
    val longitude: Double
)

@Serializable
data class VerifyRequest(
    val value: Int
)

@Serializable
data class VerifyResponse(
    val message: String,
    val vote_type: String? = null,
    val report: Report
)

@Serializable
data class VoteStatusResponse(
    val has_voted: Boolean,
    val vote_type: String? = null
)
